judge/models/tests/test_contest.py

```diff
@@ -42,14 +51,61 @@ def setUpTestData(self):
             start_time=_now - timezone.timedelta(days=1),
             end_time=_now + timezone.timedelta(days=100),
             is_visible=True,
-            hide_scoreboard=True,
+            scoreboard_visibility=Contest.SCOREBOARD_AFTER_CONTEST,
             problem_label_script='''
                 function(n)
                     return tostring(math.floor(n))
                 end
             ''',
         )
 
+        self.contest_hidden_scoreboard_contest = create_contest(
+            key='contest_scoreboard',
+            start_time=_now - timezone.timedelta(days=10),
+            end_time=_now + timezone.timedelta(days=100),
+            time_limit=timezone.timedelta(days=1),
+            is_visible=True,
+            scoreboard_visibility=Contest.SCOREBOARD_AFTER_CONTEST,
+        )
+
+        self.particip_hidden_scoreboard_contest = create_contest(
+            key='particip_scoreboard',
+            start_time=_now - timezone.timedelta(days=10),
+            end_time=_now + timezone.timedelta(days=100),
+            time_limit=timezone.timedelta(days=1),
+            is_visible=True,
+            scoreboard_visibility=Contest.SCOREBOARD_AFTER_PARTICIPATION,
+        )
+
+        self.visible_scoreboard_contest = create_contest(
+            key='visible_scoreboard',
+            start_time=_now - timezone.timedelta(days=10),
+            end_time=_now + timezone.timedelta(days=100),
+            time_limit=timezone.timedelta(days=1),
+            is_visible=True,
+            scoreboard_visibility=Contest.SCOREBOARD_VISIBLE,
+        )
+
+        for contest_key in ('contest_scoreboard', 'particip_scoreboard', 'visible_scoreboard'):
+            create_contest_participation(
+                contest=contest_key,
+                user='normal_before_window',
+            ).save()
+
+            create_contest_participation(
+                contest=contest_key,
+                user='normal_during_window',
+                real_start=_now - timezone.timedelta(hours=1),
+                virtual=ContestParticipation.LIVE,
```

Comment:
You should also add a test for `particip_scoreboard` and `virtual = 1` (i.e. not live participation) to ensure it handles that case correctly.

Reply:
I'm assuming `virtual = -1` for spectate?